pkg/fetch/fetching_fetcher.go

```diff
@@ -0,0 +1,232 @@
+)
+
+type fetchingFetcher struct {
+	contentAddressableStorage blobstore.BlobAccess
+	executionClient           remoteexecution.ExecutionClient
+	maximumMessageSizeBytes   int
+}
+
+// NewFetchingFetcher creates a new Fetcher that is capable of itself
+// fetching resources from other places (as defined in the
+// qualifier_translator).
+func NewFetchingFetcher(contentAddressableStorage blobstore.BlobAccess, client grpc.ClientConnInterface, maximumMessageSizeBytes int) Fetcher {
+	return &fetchingFetcher{
+		contentAddressableStorage: contentAddressableStorage,
+		executionClient:           remoteexecution.NewExecutionClient(client),
+		maximumMessageSizeBytes:   maximumMessageSizeBytes,
+	}
+}
+
+func (ff *fetchingFetcher) fetchCommon(ctx context.Context, req *remoteasset.FetchBlobRequest) (*remoteexecution.ActionResult, string, string, error) {
+	instanceName, err := digest.NewInstanceName(req.InstanceName)
+	if err != nil {
+		return nil, "", "", err
+	}
+	commandGenerator, err := qualifier.QualifiersToCommand(req.Qualifiers)
+	if err != nil {
+		return nil, "", "", err
+	}
+	for _, uri := range req.Uris {
+		command := commandGenerator(uri)
+		commandDigest, err := storage.ProtoToDigest(command)
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		action := &remoteexecution.Action{
+			CommandDigest:   commandDigest,
+			InputRootDigest: storage.EmptyDigest,
+		}
+		actionDigest, err := storage.ProtoToDigest(action)
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		actionPb, err := proto.Marshal(action)
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		commandPb, err := proto.Marshal(command)
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		bbActionDigest, err := instanceName.NewDigestFromProto(actionDigest)
+		if err != nil {
+			return nil, "", "", err
+		}
+		err = ff.contentAddressableStorage.Put(ctx, bbActionDigest, buffer.NewCASBufferFromByteSlice(bbActionDigest, actionPb, buffer.UserProvided))
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		bbCommandDigest, err := instanceName.NewDigestFromProto(commandDigest)
+		if err != nil {
+			return nil, "", "", err
+		}
+		err = ff.contentAddressableStorage.Put(ctx, bbCommandDigest, buffer.NewCASBufferFromByteSlice(bbCommandDigest, commandPb, buffer.UserProvided))
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		stream, err := ff.executionClient.Execute(ctx, &remoteexecution.ExecuteRequest{
+			InstanceName: req.InstanceName,
+			ActionDigest: actionDigest,
+		})
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		response := &remoteexecution.ExecuteResponse{}
+		for {
+			operation, err := stream.Recv()
+			if err != nil {
+				return nil, "", "", err
+			}
+			if operation.GetDone() {
+				err = ptypes.UnmarshalAny(operation.GetResponse(), response)
+				if err != nil {
+					return nil, "", "", err
+				}
+				break
+			}
+		}
+
+		actionResult := response.GetResult()
+		if exitCode := actionResult.GetExitCode(); exitCode != 0 {
+			continue
```

Comment:
This seems like it would be a more sensible place to log something from the stderr here (I added a response below with my thoughts on adding it in the place mentioned in the review). I don't know if it feels sensible to fill up the logs of the server with failed action stderrs though.

Reply:
These are usually mirror URLs so it's not a failure case if one fails. I'm fine with leaving this unlogged if we're concerned about flooding stderr.